WebKit should track regions for content which is not scrolled by the main scroll view

mwyrzykowski opened the pull request:
#### 3723ab78e840d398c2c74c7758f7fbfc9b1477a4
<pre>
WebKit should track regions for content which is not scrolled by the main scroll view
<a href="https://bugs.webkit.org/show_bug.cgi?id=242406">https://bugs.webkit.org/show_bug.cgi?id=242406</a>

Reviewed by Tim Horton.

Compute a set of fixed rectangles which are not scrolled by
the main scroll view.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _didCommitLayerTree:]):
Update the overlay regions.

(addOverlayEventRegions):
Recursive function which builds overlay regions from the child layers.

(-[WKWebView _updateOverlayRegions:destroyedLayers:]):
If there are destroyed or changed layers, recompute the set of overlay
regions and cleanup any removed layers which may exist in the fixed IDs
map as well as the overlay region map.

(-[WKWebView _overlayRegions]):
Compute the overlay regions as an array of NSData instances.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::fixedScrollingNodeLayerIDs const):
Return a reference to the fixed node IDs.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxy::removeFixedScrollingNodeLayerIDs):
Remove layerIDs for layers which no longer exist.

(WebKit::RemoteScrollingCoordinatorProxy::connectStateNodeLayers):
Save the fixed node IDs.

* Source/WebKit/UIProcess/ios/WKScrollView.mm:
(-[WKScrollView overlayRegions]):
Call into the web view to get the overlay regions.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
Add mutator for WKWebView to update the overlay regions.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
(WebKit::RemoteLayerTreeHost::overlayRegionsWithIDs const):
Add hash map for tracking overlay regions.

(WebKit::RemoteLayerTreeHost::updateOverlayRegionsWithIDs):
Allow updating the hash map.

Canonical link: <a href="https://commits.webkit.org/252477@main">https://commits.webkit.org/252477@main</a>
</pre>

hortont424 commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
```diff
@@ -292,8 +293,12 @@ static void updateCustomAppearance(CALayer *layer, GraphicsLayer::CustomAppearan
-    if (properties.changedProperties & RemoteLayerTreeTransaction::EventRegionChanged)
+    if (properties.changedProperties & RemoteLayerTreeTransaction::EventRegionChanged) {
         updateLayersForInteractionRegions(layer, properties);
+#if ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION)
```
I think you probably want to move the `ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)` inside the condition, only guarding updateLayersForInteractionRegions, like yours

mwyrzykowski replied:
and should the whole block be guarded by `#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION) || ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION)` ? Otherwise it could be dead code as it would have no side effects, right?

smfr replied:
Just duplicate the `if (properties.changedProperties & RemoteLayerTreeTransaction::EventRegionChanged)` check to simplify the #ifdefs. It's a bit check; super cheap.

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeOverlayRegionLayers.mm:
```diff
@@ -0,0 +1,163 @@
+
+        [exclusionRegionLayer setBackgroundColor:cachedCGColor({ WebCore::SRGBA<float>(1, 0, 0, .3) }).get()];
+
+        [exclusionRegionLayer.get() setValue:@1 forKey:debugRenderKey];
```
no .get() here.

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeOverlayRegionLayers.mm:
```diff
@@ -0,0 +1,163 @@
+{
+    static void* internalOverlayRegionsKey = &internalOverlayRegionsKey;
+
+    NSMapTable *mapTable = objc_getAssociatedObject(self, internalOverlayRegionsKey);
```
Since WKScrollView is *ours*, you could just add an internal property instead of using associated objects here, no?

smfr replied:
I would prefer this WKScrollView code to exist in WKScrollView.mm

mwyrzykowski replied:
Sounds good, that is easier than using associated objects too

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeOverlayRegionLayers.mm:
```diff
@@ -0,0 +1,163 @@
+            if (auto scrollContainerID = node->actingScrollContainerID()) {
+                RemoteLayerTreeNode *scollingContainerNode = layerTreeHost.nodeForID(scrollContainerID);
+                CALayer *scrollLayer = scollingContainerNode ? scollingContainerNode->uiView().layer : nil;
+                if (auto value = (NSNumber*)[scrollLayer valueForKey:fixedScrollableNodeLayerKey])
```
Lots of places your ObjC stars are on the wrong side :) (NSNumber *)

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeOverlayRegionLayers.mm:
```diff
@@ -0,0 +1,163 @@
+            }
+        }
+
+        if (auto value = (NSNumber*)[layer valueForKey:fixedScrollableNodeLayerKey])
```
I wonder if you want boolForKey/setBool:forKey: in a variety of places here

mwyrzykowski replied:
oh I didn't realize NSKeyValueCoding had `boolForKey`. When I try it, I see:
`error: 'CALayer' may not respond to 'boolForKey:'`

is there an header somewhere for CALayer for `boolForKey`?

hortont424 replied:
Oh, perhaps you're right! I mixed up the NSUserDefaults helpers with KVC

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
```diff
@@ -77,8 +79,11 @@
+        if (currNode->hasChangedProperty(ScrollingStateNode::Property::Layer)) {
+            CALayer *layer = layerTreeHost.layerForID(PlatformLayerID { currNode->layer() });
+            currNode->setLayer(layer);
+            [layer setValue:(currNode->isFixedNode() ? @1 : @0) forKey:fixedScrollableNodeLayerKey];
```
This should be wrapped in `#if ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION)`: `setValue:` is not cheap.

smfr replied:
Also not really a fan of stuffing values onto layers. Maybe we should propagate the ScrollingNodeID through to RemoteLayerTreeNode so we can get to the ScrollingTreeNode from a layer.

mwyrzykowski replied:
I filed https://bugs.webkit.org/show_bug.cgi?id=242471 to track this issue

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeOverlayRegionLayers.mm:
```diff
@@ -0,0 +1,163 @@
+
+void updateOverlayRectForNonScrollableRegion(CALayer *layer, const RemoteLayerTreeTransaction::LayerProperties& properties, const RemoteLayerTreeHost& layerTreeHost)
+{
+    NSString * const debugRenderKey = @"fixedPositionLayerDebugRenderKey";
```
I think you could delete a *lot* of code if you deleted the debugging layers... which might simplify things a lot. The code you borrowed this from actually makes layers for reporting information, and the debugging is incidental, but all of this *just* for debugging seems a bit overkill.

smfr replied:
Agree. Or at least separate out the debugging code into separate functions so it doesn't clutter the logic of the production code.

mwyrzykowski replied:
Production logic has been separated from the debugging function

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeOverlayRegionLayers.mm:
```diff
@@ -0,0 +1,163 @@
+{
+    extern NSString *fixedScrollableNodeLayerKey;
+
+    for ( ; view; view = [view superview]) {
```
Weird to (potentially) march up the entire view hierarchy, potentially through the views of some WKWebView embedder.

mwyrzykowski replied:
I got the idea from `static bool isScrolledBy(WKChildScrollView* scrollView, UIView *hitView)` in RemoteLayerTreeViews.mm but I'm certainly open to alternatives

whsieh commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeOverlayRegionLayers.mm:
```diff
@@ -0,0 +1,163 @@
+        }
+
+        if (auto value = (NSNumber*)[layer valueForKey:fixedScrollableNodeLayerKey])
+            return value.intValue ? false : true;
```
Nit - `!value.intValue`?

whsieh commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeOverlayRegionLayers.mm:
```diff
@@ -0,0 +1,163 @@
+                RemoteLayerTreeNode *scollingContainerNode = layerTreeHost.nodeForID(scrollContainerID);
+                CALayer *scrollLayer = scollingContainerNode ? scollingContainerNode->uiView().layer : nil;
+                if (auto value = (NSNumber*)[scrollLayer valueForKey:fixedScrollableNodeLayerKey])
+                    return value.intValue ? false : true;
```
Ditto, re: `!value.intValue`

whsieh commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeOverlayRegionLayers.mm:
```diff
@@ -0,0 +1,163 @@
+#import "RemoteLayerTreeHost.h"
+#import "WKScrollView.h"
+#import "WKWebViewIOS.h"
+#import <QuartzCore/QuartzCore.h>
```
Nit - you can avoid this import, since we're importing the SPI header below.

mwyrzykowski replied:
Oh I don't think I need the SPI header, maybe I should remove that one?

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeOverlayRegionLayers.mm:
```diff
@@ -0,0 +1,163 @@
+        if (auto value = (NSNumber*)[layer valueForKey:fixedScrollableNodeLayerKey])
+            return value.intValue ? false : true;
+
+        for (CALayer *sublayer in layer.sublayers) {
```
I'm not sure what's going on here. The outer loop walks up the view hierarchy, but there's a sublayer traversal here?

mwyrzykowski replied:
I'll investigate this and confirm. I found the ScrollingStateFixedNode's layer to be a sublayer of a view's layer from the RemoteLayerTreeNode. Maybe the view hierarchy transversal is not even necessary.

mwyrzykowski replied:
I debugged more and I find the ScrollingStateFixedNode's layer on the node's second superview's layer's third sublayer, i.e.,:
```
[(CALayer*)(((UIView*)[[nodeView superview] superview]).layer.sublayers[2]) valueForKey:@"WKFixedScrollableNodeLayerKey"]	NSConstantIntegerNumber *	0x283c88088	0x0000000283c88088
NSNumber	NSNumber	
_encoding	char *const	"i"	0x000000028307d7cc
_value	long long	1
```
where `nodeView` is the one in `updateDebugOverlayRectForView`

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeOverlayRegionLayers.mm:
```diff
@@ -0,0 +1,163 @@
+    if (![mainScrollView isKindOfClass:WKScrollView.class])
+        return;
+
+    [mainScrollView willChangeValueForKey:@"overlayRegions"];
```
Would be nice to avoid the will/didChange if the rects didn't actually change.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
```diff
@@ -129,6 +133,9 @@ class RemoteScrollingCoordinatorProxy {
     bool m_propagatesMainFrameScrolls { true };
     HashSet<WebCore::GraphicsLayer::PlatformLayerID> m_layersWithScrollingRelations;
+#if ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION)
+    HashSet<WebCore::GraphicsLayer::PlatformLayerID> m_scrollingNodeFixedLayerIDs;
```
Nothing removes entries from this HashSet when layers are destroyed, so it will grow larger and larger. We do send the list of removed layers to the UI process, so you could remove them using that.

mwyrzykowski replied:
Are those tracked by `RemoteLayerTreeTransaction::destroyedLayers()` or should I be getting the destroyed list from elsewhere? I tried iterating over that list in `_didCommitLayerTree` but the list of destroyed layer IDs doesn't seem to match up with the scrolling ID nodes so they weren't getting removed. In an update I will propose checking the layer which resolves the issue but I'm not sure if we are concerned of the inefficiencies given the number of fixed scrolling nodes. However I will limit it to when there are layer tree updates which maybe is less frequent than every time `RemoteScrollingCoordinatorProxy::connectStateNodeLayers` is called?

smfr commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1038,7 +1038,41 @@ - (void)_didCommitLayerTree:(const WebKit::RemoteLayerTreeTransaction&)layerTree
     }
+
+#if ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION)
+    [self _updateOverlayRegions];
```
It would be nice to avoid this for commits where there were no changes to the layer tree, or the scrolling tree.

mwyrzykowski replied:
Is it the case that changes to the layer tree always happen after the respective scrolling tree changes? I am wondering if it is sufficient to test for `layerTreeTransaction.changedLayerProperties().size()` and skip the update if the size is 0?

smfr commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1038,7 +1038,41 @@ - (void)_didCommitLayerTree:(const WebKit::RemoteLayerTreeTransaction&)layerTree
+            continue;
+
+        auto cgRect = (CGRect)node->eventRegion().region().bounds();
+        if (CGRectIsEmpty(cgRect))
```
If the event region is empty, it probably means that the content isn't hit-testable (e.g. `pointer-events none`). Do we want to fall back to layout bounds in this case?

mwyrzykowski replied:
I think so. Or some fallback, since otherwise the fixed banner on apple.com does not generate the expected rectangle

mwyrzykowski commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1038,7 +1038,54 @@ - (void)_didCommitLayerTree:(const WebKit::RemoteLayerTreeTransaction&)layerTree
+            }
+        }
+
+        overlayRegionsWithIDs.add(layerID, rect);
```
should be `set` not `add`

mwyrzykowski commented on Source/WebKit/UIProcess/ios/WKScrollView.mm:
```diff
@@ -548,6 +555,24 @@ - (CGPoint)_puic_contentOffsetForCrownInputSequencerOffset:(double)sequencerOffs
+{
+    NSMutableArray<NSData *> *overlayRegions = [NSMutableArray array];
+    for (const auto& [layerID, rect] : _overlayRegionsWithIDs) {
+        if (!CGRectIsEmpty(rect))
```
I don't think we ever add empty rectangles, so we could probably remove the above line

mwyrzykowski commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1038,7 +1058,43 @@ - (void)_didCommitLayerTree:(const WebKit::RemoteLayerTreeTransaction&)layerTree
+        return !layerTreeHost.layerForID(it.key);
+    });
+
+    bool didChangeOverlayRegions = overlayRegionsWithIDs.size() != existingOverlayRegionsWithIDsSize;
```
`bool didChangeOverlayRegions = ` can be assigned to the result of `removeIf` instead from what I understand looking at the implementation of `removeIf`

mwyrzykowski commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1038,7 +1058,43 @@ - (void)_didCommitLayerTree:(const WebKit::RemoteLayerTreeTransaction&)layerTree
+    auto existingOverlayRegionsWithIDsSize = overlayRegionsWithIDs.size();
+
+    overlayRegionsWithIDs.removeIf([&layerTreeHost](const auto& it) {
+        return !layerTreeHost.layerForID(it.key);
```
This removal should be done after `fixedIDs` is pruned otherwise `overlayRegionsWithIDs` can contains layers which are no longer fixed.

mwyrzykowski commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
```diff
@@ -77,8 +77,14 @@
     for (auto& currNode : stateTree.nodeMap().values()) {
-        if (currNode->hasChangedProperty(ScrollingStateNode::Property::Layer))
-            currNode->setLayer(layerTreeHost.layerForID(PlatformLayerID { currNode->layer() }));
+        auto platformLayerID = PlatformLayerID { currNode->layer() };
```
this should be inside the `if` branch

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
```diff
@@ -111,6 +111,10 @@ class RemoteScrollingCoordinatorProxy {
     WebCore::ScrollingTreeScrollingNode* rootNode() const;
 
+#if ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION)
+    HashSet<WebCore::GraphicsLayer::PlatformLayerID>& fixedScrollingNodeLayerIDs() { return m_fixedScrollingNodeLayerIDs; }
```
return a const ref and make the function const.

smfr replied:
Actually, you mutate this HashSet outside the class, which is bad layering. Please keep it internal, adding a "remove nodes with IDs" function if you need to.

mwyrzykowski replied:
will do, that is why it is non-const, but I will add the removal function instead

smfr commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -105,6 +105,27 @@ static int32_t deviceOrientationForUIInterfaceOrientation(UIInterfaceOrientation
 }
 
+#if ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION)
+static void addLayerEventRegions(auto layerID, auto fixedLayerID, const auto& changedLayerProperties, auto& overlayRegionsWithIDs, bool& didChangeOverlayRegions)
```
We don't usually use auto for parameter types. Here, it think it hurts readability.

smfr commented on Source/WebKit/UIProcess/ios/WKScrollView.h:
```diff
@@ -47,6 +47,10 @@
 #endif
 
+#if ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION)
+- (HashMap<std::pair<uint64_t, uint64_t>, CGRect>&)overlayRegionsWithIDs;
```
The std::pair<uint64_t, uint64_t> is very cryptic here.

smfr replied:
Most of your code lives on WKWebView, so maybe this HashMap can live on WKWebView's internal data, with the correct types.

smfr commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -105,6 +105,27 @@ static int32_t deviceOrientationForUIInterfaceOrientation(UIInterfaceOrientation
+    }
+
+    for (auto childLayerID : layerProperties.children)
+        addLayerEventRegions(childLayerID, fixedLayerID, changedLayerProperties, overlayRegionsWithIDs, didChangeOverlayRegions);
```
The child layer is going to have its own `layerProperties` and yet you pass `changedLayerProperties` when recursing on children, so `layerProperties.changedProperties` will be consulting the wrong layer properties.

Also, I'm not sure we guarantee that LayerProperties contains useful data for any property that hasn't changed in a given transaction; you should think of LayerProperties as a set of deltas. It would be more valid to walk CALayer children once the transaction has been applied.

mwyrzykowski replied:
changedLayerProperties  is actually the entire map of type `WebKit::RemoteLayerTreeTransaction::LayerPropertiesMap` but maybe that is confusing with `auto` type here

mwyrzykowski replied:
Does it make sense using the set of deltas given `changedLayerProperties` is the entire map of deltas? The deltas is what I am interested in using to update the set of rectangles

smfr commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -1038,7 +1059,39 @@ - (void)_didCommitLayerTree:(const WebKit::RemoteLayerTreeTransaction&)layerTree
+    const auto& layerTreeHost = downcast<WebKit::RemoteLayerTreeDrawingAreaProxy>(*_page->drawingArea()).remoteLayerTreeHost();
+    auto overlayRegionsWithIDs = [_scrollView overlayRegionsWithIDs];
+    auto& fixedIDs = _page->scrollingCoordinatorProxy()->fixedScrollingNodeLayerIDs();
+    fixedIDs.removeIf([&layerTreeHost](auto layerID) {
```
Why not use destroyedLayers here? Also, what removes layers whose style changes from `position:fixed` to something else?

mwyrzykowski replied:
I tried iterating over `destroyedLayers` but they didn't match the layerIDs of the fixed nodes, which led to the HashSet growing unbounded across navigations. Do only the parent layers which are destroyed end up in destroyed layers?

mwyrzykowski commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
```diff
@@ -76,6 +76,10 @@ class RemoteLayerTreeHost {
 
     bool replayCGDisplayListsIntoBackingStore() const;
+#if ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION)
+    const HashMap<std::pair<WebCore::GraphicsLayer::PlatformLayerID, WebCore::GraphicsLayer::PlatformLayerID>, CGRect>& overlayRegionsWithIDs() const { return m_overlayRegionsWithIDs; }
```
now that this is in the RemoteLayerTreeHost I think the key can be a layerID instead of a pair of IDs

hortont424 commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
```diff
@@ -181,6 +181,10 @@ enum class TapHandlingResult : uint8_t;
 #endif
 
+#if ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION)
+- (NSArray<NSData *> *)overlayRegions;
```
This needs a leading underscore

hortont424 commented on Source/WebKit/UIProcess/ios/WKScrollView.mm:
```diff
@@ -548,6 +548,13 @@ - (CGPoint)_puic_contentOffsetForCrownInputSequencerOffset:(double)sequencerOffs
 #endif // HAVE(PEPPER_UI_CORE)
 
+#if ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION)
+- (NSArray<NSData *> *)overlayRegions
```
This should probably also have an underscore

mwyrzykowski committed:
WebKit should track regions for content which is not scrolled by the main scroll view
https://bugs.webkit.org/show_bug.cgi?id=242406

Reviewed by Tim Horton.

Compute a set of fixed rectangles which are not scrolled by
the main scroll view.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _didCommitLayerTree:]):
Update the overlay regions.

(addOverlayEventRegions):
Recursive function which builds overlay regions from the child layers.

(-[WKWebView _updateOverlayRegions:destroyedLayers:]):
If there are destroyed or changed layers, recompute the set of overlay
regions and cleanup any removed layers which may exist in the fixed IDs
map as well as the overlay region map.

(-[WKWebView _overlayRegions]):
Compute the overlay regions as an array of NSData instances.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::fixedScrollingNodeLayerIDs const):
Return a reference to the fixed node IDs.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxy::removeFixedScrollingNodeLayerIDs):
Remove layerIDs for layers which no longer exist.

(WebKit::RemoteScrollingCoordinatorProxy::connectStateNodeLayers):
Save the fixed node IDs.

* Source/WebKit/UIProcess/ios/WKScrollView.mm:
(-[WKScrollView overlayRegions]):
Call into the web view to get the overlay regions.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
Add mutator for WKWebView to update the overlay regions.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
(WebKit::RemoteLayerTreeHost::overlayRegionsWithIDs const):
Add hash map for tracking overlay regions.

(WebKit::RemoteLayerTreeHost::updateOverlayRegionsWithIDs):
Allow updating the hash map.

Canonical link: https://commits.webkit.org/252477@main
